lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -80,6 +89,31 @@ static void addCanonicalizerMethod(OwningRewritePatternList &results,
   results.insert<Folder>(context);
 }
 
+/// Performs constant folding `calculate` with element-wise behavior on the two
+/// attributes in `operands` and returns the result if possible, with the
+/// specified result type.
+///
+/// TODO: Maybe the upstream constFoldBinaryOp can be made to have a
+/// configurable result type as well.
+template <class AttrElementT,
+          class ElementValueT = typename AttrElementT::ValueType,
+          class CalculationT = function_ref<bool(ElementValueT, ElementValueT)>>
+Attribute constFoldBinaryRelationalOp(ArrayRef<Attribute> operands,
+                                      const CalculationT &calculate) {
+  assert(operands.size() == 2 && "binary op takes two operands");
+  if (!operands[0] || !operands[1])
+    return {};
+  if (operands[0].getType() != operands[1].getType())
+    return {};
+  if (operands[0].isa<AttrElementT>() && operands[1].isa<AttrElementT>()) {
+    auto lhs = operands[0].cast<AttrElementT>();
+    auto rhs = operands[1].cast<AttrElementT>();
```

Comment:
I think this needs to handle sign/zero extension of the attributes, xref prithian's patch: https://github.com/llvm/circt/pull/827

We need to make sure to align with the logic in LowerToRTL.

I like your approach here though, it keeps the constant folding logic tidy at the use site.

Reply:
Huh that is an excellent point. Completely forgot about `APInt`s not operating on different bit widths. That would also imply that the existing calls to `constFoldBinaryOp`, for example for `XorPrimOp` above, need to be changed to handle different bit widths as well. I believe this might crash at the moment if bit widths disagree.

Reply:
Yeah likely so

Reply:
Correction: It doesn't crash because the `constFoldBinaryOp` internally asserts that the types of the operands are identical, which also ensures that the `APInt`s match.